src/fu-security-attr.c

```diff
@@ -301,17 +305,269 @@ fu_security_attrs_to_json(FuSecurityAttrs *attrs, JsonBuilder *builder)
 {
 	g_autoptr(GPtrArray) items = NULL;
 	g_autoptr(GError) error = NULL;
-
 	json_builder_begin_object(builder);
 	json_builder_set_member_name(builder, "SecurityAttributes");
-	json_builder_begin_array(builder);
+	json_builder_begin_object(builder);
 	items = fu_security_attrs_get_all(attrs);
 	for (guint i = 0; i < items->len; i++) {
 		FwupdSecurityAttr *attr = g_ptr_array_index(items, i);
+		json_builder_set_member_name(builder, fwupd_security_attr_get_appstream_id(attr));
 		json_builder_begin_object(builder);
 		fwupd_security_attr_to_json(attr, builder);
 		json_builder_end_object(builder);
 	}
-	json_builder_end_array(builder);
 	json_builder_end_object(builder);
+	json_builder_end_object(builder);
+}
+
+guint
+fu_security_attrs_compare_hsi_score(const guint previous_hsi, const guint current_hsi)
+{
+	if (current_hsi > previous_hsi)
+		return 1;
+	else if (current_hsi < previous_hsi)
+		return -1;
+	else
+		return 0;
+}
+
+static void
+fu_security_attr_dup_json_array_to_builder(JsonBuilder *builder,
+					   JsonArray *src,
+					   const gchar *item_name)
+{
+	JsonNode *json_node;
+	if (src != NULL) {
+		json_builder_set_member_name(builder, item_name);
+		json_builder_begin_array(builder);
+		for (guint i = 0; i < json_array_get_length(src); i++) {
+			json_node = json_array_dup_element(src, i);
+			json_builder_add_value(builder, json_node);
+		}
+		json_builder_end_array(builder);
+	}
+}
+
+static void
+fu_security_attr_dup_json(JsonObject *src, JsonBuilder *builder)
+{
+	JsonArray *array_items = NULL;
+	json_builder_set_member_name(builder, FWUPD_RESULT_KEY_HSI_LEVEL);
+	json_builder_add_int_value(builder,
+				   json_object_get_int_member(src, FWUPD_RESULT_KEY_HSI_LEVEL));
+	json_builder_set_member_name(builder, FWUPD_RESULT_KEY_HSI_RESULT);
+	json_builder_add_string_value(
+	    builder,
+	    json_object_get_string_member(src, FWUPD_RESULT_KEY_HSI_RESULT));
+	json_builder_set_member_name(builder, FWUPD_RESULT_KEY_NAME);
+	json_builder_add_string_value(builder,
+				      json_object_get_string_member(src, FWUPD_RESULT_KEY_NAME));
+	if (json_object_has_member(src, FWUPD_RESULT_KEY_FLAGS) == TRUE) {
+		array_items = json_object_get_array_member(src, FWUPD_RESULT_KEY_FLAGS);
+		fu_security_attr_dup_json_array_to_builder(builder,
+							   array_items,
+							   FWUPD_RESULT_KEY_FLAGS);
+	}
+}
+
+/**
+ * fu_security_attr_flag_to_string_array:
+ *
+ * @self: a single FwupdSecurityAttr
+ *
+ * Convert flags to string and store them in GPtrArray.
+ *
+ * Since: 1.7.0
+ *
+ */
+static GPtrArray *
+fu_security_attr_flag_to_string_array(FwupdSecurityAttr *attr)
+{
+	GPtrArray *flag_array = NULL;
+	FwupdSecurityAttrFlags flags = fwupd_security_attr_get_flags(attr);
+	if (flags != FWUPD_SECURITY_ATTR_FLAG_NONE) {
+		flag_array = g_ptr_array_new();
```

Comment:
this needs to be `g_ptr_array_new_with_free_func(g_free)` to deallocate the stuff you're g_strdup'ing.

Reply:
Thanks for reminding me this. I have corrected this.